File: src/ai_usage_bar/auth/codex_auth.py
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any

from ai_usage_bar.auth import pick_str
from ai_usage_bar.models import jwt_payload

log = logging.getLogger(__name__)

CODEX_CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"
TOKEN_URL = "https://auth.openai.com/oauth/token"


def auth_json_path() -> Path:
    import os

    home = Path(os.environ.get("CODEX_HOME", Path.home() / ".codex"))
    return home / "auth.json"


def load_codex_tokens() -> dict[str, str]:
    path = auth_json_path()
    if not path.is_file():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        log.warning("Codex の auth.json を読めませんでした")
        return {}
    if not isinstance(data, dict):
        return {}

    nested = data.get("tokens")
    source: dict[str, Any] = nested if isinstance(nested, dict) else data
    access = pick_str(source, "access_token", "accessToken")
    refresh = pick_str(source, "refresh_token", "refreshToken")
    account = pick_str(source, "account_id", "accountId")
    if not account and access:
        payload = jwt_payload(access)
        auth_claim = payload.get("https://api.openai.com/auth")
        if isinstance(auth_claim, dict):
            account = pick_str(auth_claim, "chatgpt_account_id", "account_id")
        if not account:
            account = pick_str(payload, "chatgpt_account_id", "account_id")
    result: dict[str, str] = {}
    if access:
        result["access_token"] = access
    if refresh:
        result["refresh_token"] = refresh
    if account:
        result["account_id"] = account
    return result


def refresh_access_token(refresh_token: str) -> str | None:
    import httpx

    try:
        response = httpx.post(
            TOKEN_URL,
            data={
                "grant_type": "refresh_token",
                "client_id": CODEX_CLIENT_ID,
                "refresh_token": refresh_token,
                "scope": "openid profile email",
            },
            headers={"Accept": "application/json"},
            timeout=20.0,
        )
    except httpx.HTTPError:
        log.warning("Codex のトークン更新リクエストに失敗しました")
        return None
    if response.status_code >= 400:
        log.warning("Codex のトークン更新が拒否されました status=%s", response.status_code)
        return None
    try:
        payload = response.json()
    except ValueError:
        return None
    if not isinstance(payload, dict):
        return None
    token = pick_str(payload, "access_token")
    return token

File: src/AiUsageBar.Core/Auth/CodexAuth.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace AiUsageBar.Core.Auth;

public sealed record CodexRefreshResult(string AccessToken, string? RefreshToken, string? IdToken);

public static class CodexAuth
{
    public const string ClientId = "app_EMoamEEZ73f0CkXaXp7hrann";
    public const string TokenUrl = "https://auth.openai.com/oauth/token";

    public static string AuthJsonPath
    {
        get
        {
            var home = Environment.GetEnvironmentVariable("CODEX_HOME")
                ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".codex");
            return Path.Combine(home, "auth.json");
        }
    }

    public static string RefreshRequestJson(string refreshToken) =>
        JsonSerializer.Serialize(new Dictionary<string, string>
        {
            ["client_id"] = ClientId,
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refreshToken,
        });

    public static Dictionary<string, string> LoadTokens()
    {
        var path = AuthJsonPath;
        if (!File.Exists(path))
        {
            return [];
        }

        try
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
            {
                return [];
            }

            var root = doc.RootElement;
            var source = root.TryGetProperty("tokens", out var nested) && nested.ValueKind == JsonValueKind.Object
                ? nested
                : root;
            var access = Formatting.PickStr(source, "access_token", "accessToken");
            var refresh = Formatting.PickStr(source, "refresh_token", "refreshToken");
            var idToken = Formatting.PickStr(source, "id_token", "idToken");
            var account = Formatting.PickStr(source, "account_id", "accountId")
                ?? AccountIdFromJwt(idToken)
                ?? AccountIdFromJwt(access);

            var result = new Dictionary<string, string>(StringComparer.Ordinal);
            if (access is not null)
            {
                result["access_token"] = access;
            }

            if (refresh is not null)
            {
                result["refresh_token"] = refresh;
            }

            if (account is not null)
            {
                result["account_id"] = account;
            }

            return result;
        }
        catch (Exception ex) when (ex is IOException or JsonException)
        {
            AppLog.Warning("Codex の auth.json を読めませんでした");
            return [];
        }
    }

    public static bool AccessTokenNeedsRefresh(string accessToken, DateTimeOffset? now = null)
    {
        var payload = Formatting.JwtPayload(accessToken);
        if (!payload.TryGetValue("exp", out var expEl) || Formatting.Number(expEl) is not { } exp)
        {
            return false;
        }

        now ??= Formatting.UtcNow();
        try
        {
            return DateTimeOffset.FromUnixTimeSeconds((long)exp) <= now.Value.AddMinutes(2);
        }
        catch (ArgumentOutOfRangeException)
        {
            return false;
        }
    }

    public static string? AccountIdFromJwt(string? token)
    {
        if (string.IsNullOrEmpty(token))
        {
            return null;
        }

        var payload = Formatting.JwtPayload(token);
        if (payload.TryGetValue("https://api.openai.com/auth", out var claim) && claim.ValueKind == JsonValueKind.Object)
        {
            return Formatting.PickStr(claim, "chatgpt_account_id", "account_id");
        }

        return Formatting.PickStr(payload, "chatgpt_account_id", "account_id");
    }

    public static async Task<CodexRefreshResult?> RefreshAccessTokenAsync(string refreshToken, HttpClient http, CancellationToken cancellationToken)
    {
        try
        {
            using var content = new StringContent(RefreshRequestJson(refreshToken), Encoding.UTF8);
            content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
            using var request = new HttpRequestMessage(HttpMethod.Post, TokenUrl) { Content = content };
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            using var response = await http.SendAsync(request, cancellationToken).ConfigureAwait(false);
            if ((int)response.StatusCode >= 400)
            {
                AppLog.Warning($"Codex のトークン更新が拒否されました status={(int)response.StatusCode}");
                return null;
            }

            using var doc = JsonDocument.Parse(await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false));
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
            {
                return null;
            }

            var access = Formatting.PickStr(doc.RootElement, "access_token");
            if (string.IsNullOrEmpty(access))
            {
                return null;
            }

            return new CodexRefreshResult(
                access,
                Formatting.PickStr(doc.RootElement, "refresh_token"),
                Formatting.PickStr(doc.RootElement, "id_token"));
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or JsonException)
        {
            AppLog.Warning("Codex のトークン更新リクエストに失敗しました");
            return null;
        }
    }

    public static void ApplyRefreshedTokens(CodexRefreshResult tokens, string? path = null)
    {
        path ??= AuthJsonPath;
        if (!File.Exists(path))
        {
            return;
        }

        try
        {
            var root = JsonNode.Parse(File.ReadAllText(path)) as JsonObject ?? [];
            var nested = root["tokens"] as JsonObject ?? [];
            nested["access_token"] = tokens.AccessToken;
            if (!string.IsNullOrEmpty(tokens.RefreshToken))
            {
                nested["refresh_token"] = tokens.RefreshToken;
            }

            if (!string.IsNullOrEmpty(tokens.IdToken))
            {
                nested["id_token"] = tokens.IdToken;
            }

            var account = AccountIdFromJwt(tokens.AccessToken) ?? AccountIdFromJwt(tokens.IdToken);
            if (!string.IsNullOrEmpty(account))
            {
                nested["account_id"] = account;
            }

            root["tokens"] = nested;
            root["last_refresh"] = Formatting.UtcNow().ToString("o");
            File.WriteAllText(path, root.ToJsonString(new JsonSerializerOptions { WriteIndented = true }) + Environment.NewLine);
            AppLog.Info("Codex のトークンを更新して保存しました");
        }
        catch (Exception ex) when (ex is IOException or JsonException)
        {
            AppLog.Warning("Codex の auth.json へ更新トークンを書けませんでした");
        }
    }

    public static IReadOnlyList<string>? FindCli()
    {
        var names = new[] { "codex.cmd", "codex.exe", "codex.bat", "codex" };
        foreach (var dir in CliSearchDirs())
        {
            foreach (var name in names)
            {
                var candidate = Path.Combine(dir, name);
                if (!File.Exists(candidate))
                {
                    continue;
                }

                if (candidate.EndsWith(".ps1", StringComparison.OrdinalIgnoreCase))
                {
                    return ["powershell.exe", "-NoProfile", "-NonInteractive", "-File", candidate];
                }

                return [candidate];
            }
        }

        return null;
    }

    private static IEnumerable<string> CliSearchDirs()
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var dir in (Environment.GetEnvironmentVariable("PATH") ?? "").Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
        {
            var trimmed = dir.Trim().Trim('"');
            if (seen.Add(trimmed))
            {
                yield return trimmed;
            }
        }

        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var local = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        foreach (var extra in new[]
                 {
                     Path.Combine(local, "mise", "shims"),
                     Path.Combine(home, ".local", "bin"),
                     Path.Combine(home, ".cargo", "bin"),
                 })
        {
            if (seen.Add(extra))
            {
                yield return extra;
            }
        }
    }
}
